drivers/can/can_mcan.c

```diff
@@ -0,0 +1,867 @@
+		}
+	}
+#endif
+
+#ifdef CONFIG_CAN_FD_MODE
+	timing_data.sjw = cfg->sjw_data;
+	can_mcan_configure_timing(can, &timing, &timing_data);
+#else
+	timing.sjw = cfg->sjw;
+	can_mcan_configure_timing(can, &timing, NULL);
+#endif
+
+	can->ie = CAN_MCAN_IE_BO | CAN_MCAN_IE_EW | CAN_MCAN_IE_EP |
+		  CAN_MCAN_IE_MRAF | CAN_MCAN_IE_TEFL | CAN_MCAN_IE_TEFN |
+		  CAN_MCAN_IE_RF0N | CAN_MCAN_IE_RF1N | CAN_MCAN_IE_RF0L |
+		  CAN_MCAN_IE_RF1L;
+
+#ifdef CONFIG_CAN_STM32FD
+	can->ils = CAN_MCAN_ILS_RXFIFO0 | CAN_MCAN_ILS_RXFIFO1;
+#else
+	can->ils = CAN_MCAN_ILS_RF0N | CAN_MCAN_ILS_RF1N;
+#endif
+	can->ile = CAN_MCAN_ILE_EINT0 | CAN_MCAN_ILE_EINT1;
+	/* Interrupt on every TX fifo element*/
+	can->txbtie = CAN_MCAN_TXBTIE_TIE;
+
+	ret = can_leave_init_mode(can, K_MSEC(CAN_INIT_TIMEOUT));
+	if (ret) {
+		LOG_ERR("Failed to leave init mode");
+		return -EIO;
+	}
+
+	/* No memset because only aligned ptr are allowed */
+	for (uint32_t *ptr = (uint32_t *)msg_ram;
+	     ptr < (uint32_t *)msg_ram +
+		   sizeof(struct can_mcan_msg_sram) / sizeof(uint32_t);
+	     ptr++) {
+		*ptr = 0;
+	}
+
+	return 0;
+}
+
+static void can_mcan_state_change_handler(const struct can_mcan_config *cfg,
+					  struct can_mcan_data *data)
+{
+	enum can_state state;
+	struct can_bus_err_cnt err_cnt;
+
+	state = can_mcan_get_state(cfg, &err_cnt);
+
+	if (data->state_change_isr) {
+		data->state_change_isr(state, err_cnt);
+	}
+}
+
+static void can_mcan_tc_event_handler(struct can_mcan_reg *can,
+				      struct can_mcan_msg_sram *msg_ram,
+				      struct can_mcan_data *data)
+{
+	volatile struct can_mcan_tx_event_fifo *tx_event;
+	can_tx_callback_t tx_cb;
+	uint32_t event_idx, tx_idx;
+
+	while (can->txefs & CAN_MCAN_TXEFS_EFFL) {
+		event_idx = (can->txefs & CAN_MCAN_TXEFS_EFGI) >>
+			    CAN_MCAN_TXEFS_EFGI_POS;
+		tx_event = &msg_ram->tx_event_fifo[event_idx];
+		tx_idx = tx_event->mm.idx;
+		/* Acknowledge TX event */
+		can->txefa = event_idx;
+
+		k_sem_give(&data->tx_sem);
+
+		tx_cb = data->tx_fin_cb[tx_idx];
+		if (tx_cb == NULL) {
+			k_sem_give(&data->tx_fin_sem[tx_idx]);
+		} else {
+			tx_cb(CAN_TX_OK, data->tx_fin_cb_arg[tx_idx]);
+		}
+	}
+}
+
+void can_mcan_line_0_isr(const struct can_mcan_config *cfg,
+			 struct can_mcan_msg_sram *msg_ram,
+			 struct can_mcan_data *data)
+{
+	struct can_mcan_reg *can = cfg->can;
+
+	do {
+		if (can->ir & (CAN_MCAN_IR_BO | CAN_MCAN_IR_EP |
+			       CAN_MCAN_IR_EW)) {
+			can->ir = CAN_MCAN_IR_BO | CAN_MCAN_IR_EP |
+				  CAN_MCAN_IR_EW;
+			can_mcan_state_change_handler(cfg, data);
+		}
+		/* TX event FIFO new entry */
+		if (can->ir & CAN_MCAN_IR_TEFN) {
+			can->ir = CAN_MCAN_IR_TEFN;
+			can_mcan_tc_event_handler(can, msg_ram, data);
+		}
+
+		if (can->ir & CAN_MCAN_IR_TEFL) {
+			can->ir = CAN_MCAN_IR_TEFL;
+			LOG_ERR("TX FIFO element lost");
+			k_sem_give(&data->tx_sem);
+		}
+
+		if (can->ir & CAN_MCAN_IR_ARA) {
+			can->ir  = CAN_MCAN_IR_ARA;
+			LOG_ERR("Acces to reserved address");
```

Comment:
I found a small typo:
```suggestion
			LOG_ERR("Access to reserved address");
```